File: kotlin/src/main/kotlin/chromahash/Dct.kt
package chromahash

import kotlin.math.PI
import kotlin.math.abs
import kotlin.math.max

/**
 * Compute the AC coefficient scan order for an nx×ny grid keyed on aspectByte.
 * Per spec §6.2 (v0.4): coefficients sorted ascending by per-pixel frequency priority
 * `(cx*h)^2 + (cy*w)^2` where (w,h) = decodeOutputSize(aspectByte).
 * Ties broken by (cx, cy). Excludes DC at (0,0).
 */
internal fun scanOrder(
    nx: Int,
    ny: Int,
    aspectByte: Int,
): List<Pair<Int, Int>> {
    val (w, h) = decodeOutputSize(aspectByte)

    data class Entry(val priority: Long, val cx: Int, val cy: Int)
    val entries = mutableListOf<Entry>()
    for (cy in 0 until ny) {
        val cxStart = if (cy == 0) 1 else 0
        var cx = cxStart
        while (cx * ny < nx * (ny - cy)) {
            val a = cx.toLong() * h.toLong()
            val b = cy.toLong() * w.toLong()
            entries.add(Entry(a * a + b * b, cx, cy))
            cx++
        }
    }
    entries.sortWith(
        compareBy<Entry> { it.priority }.thenBy { it.cx }.thenBy { it.cy },
    )
    return entries.map { Pair(it.cx, it.cy) }
}

/**
 * Forward DCT encode for a channel. Per spec §12.6 dctEncode (v0.4).
 * AC values are emitted in `scan` order.
 * Returns Triple(dc, acCoefficients, scale).
 */
internal fun dctEncode(
    channel: DoubleArray,
    w: Int,
    h: Int,
    scan: List<Pair<Int, Int>>,
): Triple<Double, DoubleArray, Double> {
    val wh = (w * h).toDouble()

    // DC = mean (cos(0)=1 everywhere)
    var sum = 0.0
    for (v in channel) sum += v
    val dc = sum / wh

    val acList = DoubleArray(scan.size)
    var scale = 0.0

    for ((idx, pair) in scan.withIndex()) {
        val (cx, cy) = pair
        var f = 0.0
        for (y in 0 until h) {
            val fy = portableCos(PI / h.toDouble() * cy.toDouble() * (y.toDouble() + 0.5))
            for (x in 0 until w) {
                f +=
                    channel[x + y * w] *
                    portableCos(PI / w.toDouble() * cx.toDouble() * (x.toDouble() + 0.5)) *
                    fy
            }
        }
        f /= wh
        acList[idx] = f
        scale = max(scale, abs(f))
    }

    // Floor near-zero scale to exactly zero. When the channel is (near-)constant,
    // floating-point noise in cosine sums produces tiny AC values. Without this
    // threshold, dividing AC/scale amplifies platform-specific ULP differences
    // (e.g. different cbrt implementations) into divergent quantized codes.
    if (scale < 1e-10) {
        for (i in acList.indices) acList[i] = 0.0
        scale = 0.0
    }

    return Triple(dc, acList, scale)
}

/**
 * Precompute cosine table for DCT: table[freq][pos] = cos(PI/dim * freq * (pos+0.5)).
 * Per spec §12.6. Uses portableCos for cross-platform determinism.
 */
internal fun precomputeCosTable(
    dim: Int,
    maxFreq: Int,
): Array<DoubleArray> =
    Array(maxFreq) { freq ->
        DoubleArray(dim) { pos ->
            portableCos(PI / dim.toDouble() * freq.toDouble() * (pos.toDouble() + 0.5))
        }
    }

/**
 * Forward DCT encode using precomputed cosine tables. Per spec §12.6 (v0.4).
 * Semantically identical to dctEncode but avoids redundant cosine evaluations.
 * cosX/cosY must have rows for all (cx, cy) in `scan`.
 */
internal fun dctEncodeSeparable(
    channel: DoubleArray,
    w: Int,
    h: Int,
    scan: List<Pair<Int, Int>>,
    cosX: Array<DoubleArray>,
    cosY: Array<DoubleArray>,
): Triple<Double, DoubleArray, Double> {
    val wh = (w * h).toDouble()

    // DC = mean (cosX[0]/cosY[0] are all-ones by construction)
    var sum = 0.0
    for (v in channel) sum += v
    val dc = sum / wh

    val acList = DoubleArray(scan.size)
    var scale = 0.0

    for ((idx, pair) in scan.withIndex()) {
        val (cx, cy) = pair
        val cxRow = cosX[cx]
        val cyRow = cosY[cy]
        var f = 0.0
        for (y in 0 until h) {
            val fy = cyRow[y]
            for (x in 0 until w) {
                f += channel[x + y * w] * cxRow[x] * fy
            }
        }
        f /= wh
        acList[idx] = f
        scale = max(scale, abs(f))
    }

    if (scale < 1e-10) {
        for (i in acList.indices) acList[i] = 0.0
        scale = 0.0
    }

    return Triple(dc, acList, scale)
}

/** Inverse DCT at a single pixel (x, y) for a channel. */
internal fun dctDecodePixel(
    dc: Double,
    ac: DoubleArray,
    scanOrder: List<Pair<Int, Int>>,
    x: Int,
    y: Int,
    w: Int,
    h: Int,
): Double {
    var value = dc
    for ((j, pair) in scanOrder.withIndex()) {
        val (cx, cy) = pair
        val cxFactor = if (cx > 0) 2.0 else 1.0
        val cyFactor = if (cy > 0) 2.0 else 1.0
        val fx = portableCos(PI / w.toDouble() * cx.toDouble() * (x.toDouble() + 0.5))
        val fy = portableCos(PI / h.toDouble() * cy.toDouble() * (y.toDouble() + 0.5))
        value += ac[j] * fx * fy * cxFactor * cyFactor
    }
    return value
}

/**
 * Inverse DCT at a single pixel using precomputed cosine tables. Per spec §12.6.
 * Semantically identical to dctDecodePixel but reads cosX[cx][x] / cosY[cy][y].
 * The cx/cy factors stay as separate multiplies to preserve the float operation order.
 */
internal fun dctDecodePixelSeparable(
    dc: Double,
    ac: DoubleArray,
    scanOrder: List<Pair<Int, Int>>,
    x: Int,
    y: Int,
    cosX: Array<DoubleArray>,
    cosY: Array<DoubleArray>,
): Double {
    var value = dc
    for ((j, pair) in scanOrder.withIndex()) {
        val (cx, cy) = pair
        val cxFactor = if (cx > 0) 2.0 else 1.0
        val cyFactor = if (cy > 0) 2.0 else 1.0
        val fx = cosX[cx][x]
        val fy = cosY[cy][y]
        value += ac[j] * fx * fy * cxFactor * cyFactor
    }
    return value
}

File: python/chromahash/_dct.py
"""DCT encode/decode for ChromaHash channels."""

import math

from chromahash._aspect import decode_output_size
from chromahash._math_utils import portable_cos


def scan_order(nx: int, ny: int, aspect_byte: int) -> list[tuple[int, int]]:
    """Compute the AC coefficient scan order for an nx×ny grid keyed on aspect_byte.
    Per spec §6.2 (v0.4): coefficients are sorted ascending by per-pixel frequency priority
    `(cx*h)^2 + (cy*w)^2` where (w,h) = decodeOutputSize(aspect_byte).
    Ties broken by (cx, cy). Excludes DC at (0,0).
    """
    w, h = decode_output_size(aspect_byte)
    entries: list[tuple[int, int, int]] = []
    for cy in range(ny):
        cx = 1 if cy == 0 else 0
        while cx * ny < nx * (ny - cy):
            priority = (cx * h) * (cx * h) + (cy * w) * (cy * w)
            entries.append((priority, cx, cy))
            cx += 1
    entries.sort()
    return [(cx, cy) for (_, cx, cy) in entries]


def dct_encode(
    channel: list[float],
    w: int,
    h: int,
    scan: list[tuple[int, int]],
) -> tuple[float, list[float], float]:
    """Forward DCT encode for a channel. Per spec §12.6 dctEncode (v0.4).
    Returns (dc, ac_coefficients, scale). AC values are emitted in `scan` order.
    """
    wh = w * h
    dc = sum(channel) / wh

    ac: list[float] = []
    scale = 0.0
    for cx, cy in scan:
        f = 0.0
        for y in range(h):
            fy = portable_cos(math.pi / h * cy * (y + 0.5))
            for x in range(w):
                f += channel[x + y * w] * portable_cos(math.pi / w * cx * (x + 0.5)) * fy
        f /= wh
        ac.append(f)
        scale = max(scale, abs(f))

    # Floor near-zero scale to exactly zero. When the channel is (near-)constant,
    # floating-point noise produces tiny AC values. Without this threshold,
    # dividing AC/scale amplifies platform-specific ULP differences into
    # divergent quantized codes.
    if scale < 1e-10:
        ac = [0.0] * len(ac)
        scale = 0.0

    return (dc, ac, scale)


def precompute_cos_table(dim: int, max_freq: int) -> list[list[float]]:
    """Precompute cosine table for DCT: table[freq][pos] = cos(pi/dim * freq * (pos+0.5)).
    Per spec §12.6. Uses portable_cos for cross-platform determinism.
    """
    return [
        [portable_cos(math.pi / dim * freq * (pos + 0.5)) for pos in range(dim)]
        for freq in range(max_freq)
    ]


def dct_encode_separable(
    channel: list[float],
    w: int,
    h: int,
    scan: list[tuple[int, int]],
    cos_x: list[list[float]],
    cos_y: list[list[float]],
) -> tuple[float, list[float], float]:
    """Forward DCT encode using precomputed cosine tables. Per spec §12.6 (v0.4).
    Semantically identical to dct_encode but avoids redundant cosine evaluations.
    cos_x/cos_y must have rows for all (cx, cy) in `scan`.
    """
    wh = w * h
    dc = sum(channel) / wh

    ac: list[float] = []
    scale = 0.0
    for cx, cy in scan:
        cx_row = cos_x[cx]
        cy_row = cos_y[cy]
        f = 0.0
        for y in range(h):
            fy = cy_row[y]
            for x in range(w):
                f += channel[x + y * w] * cx_row[x] * fy
        f /= wh
        ac.append(f)
        scale = max(scale, abs(f))

    if scale < 1e-10:
        ac = [0.0] * len(ac)
        scale = 0.0

    return (dc, ac, scale)


def dct_decode_pixel(
    dc: float,
    ac: list[float],
    scan_order: list[tuple[int, int]],
    x: int,
    y: int,
    w: int,
    h: int,
) -> float:
    """Inverse DCT at a single pixel (x, y) for a channel."""
    value = dc
    for j, (cx, cy) in enumerate(scan_order):
        cx_factor = 2.0 if cx > 0 else 1.0
        cy_factor = 2.0 if cy > 0 else 1.0
        fx = portable_cos(math.pi / w * cx * (x + 0.5))
        fy = portable_cos(math.pi / h * cy * (y + 0.5))
        value += ac[j] * fx * fy * cx_factor * cy_factor
    return value


def dct_decode_pixel_separable(
    dc: float,
    ac: list[float],
    scan_order: list[tuple[int, int]],
    x: int,
    y: int,
    cos_x: list[list[float]],
    cos_y: list[list[float]],
) -> float:
    """Inverse DCT at a single pixel using precomputed cosine tables. Per spec §12.6.
    Semantically identical to dct_decode_pixel but reads cos_x[cx][x] / cos_y[cy][y].
    The cx/cy factors stay as separate multiplies to preserve the float operation order.
    """
    value = dc
    for j, (cx, cy) in enumerate(scan_order):
        cx_factor = 2.0 if cx > 0 else 1.0
        cy_factor = 2.0 if cy > 0 else 1.0
        fx = cos_x[cx][x]
        fy = cos_y[cy][y]
        value += ac[j] * fx * fy * cx_factor * cy_factor
    return value
